sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala

```diff
@@ -112,6 +112,23 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper with No
       } else {
         super.visitAggregateFunction(funcName, isDistinct, inputs)
       }
+
+    override def visitCast(expr: String, exprDataType: DataType, dataType: DataType): String = {
+      val databaseTypeDefinition = dataType match {
+        // MySQL uses CHAR in the cast function for the type LONGTEXT
+        case StringType => "CHAR"
+        // MySQL uses SIGNED INTEGER in the cast function for SMALLINT, INTEGER and BIGINT.
+        // To avoid breaking code relying on ResultSet metadata, we support BIGINT only at
+        // this time.
+        case LongType => "SIGNED INTEGER"
+        case ShortType | IntegerType =>
+            QueryExecutionErrors.unsupportedJdbcTypeError(dataType.catalogString)
```

Comment:
Can we add a test for not being pushed?

Reply:
Sure, let me add a test for these cases.

Reply:
The test for not supported types are covered, PTAL.

Reply:
Personally, LongType can transformed to "SIGNED INTEGER", but ShortType and IntegerType, looks weird.
@yaooqinn How about your opinion?